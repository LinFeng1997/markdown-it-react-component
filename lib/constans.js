export const babelTransformOptions = {
  presets: ['stage-3', 'react', 'es2015'],
  plugins: ['filterXSS']
};

var attr_name     = '[a-zA-Z_:][a-zA-Z0-9:._-]*';
var unquoted      = '[^"\'=<>`\\x00-\\x20]+';
var single_quoted = '\'[^\']*\'';
var double_quoted = '"[^"]*"';

var attr_value  = '(?:' + unquoted + '|' + single_quoted + '|' + double_quoted + ')';
var attribute   = '(?:\\s+' + attr_name + '(?:\\s*=\\s*' + attr_value + ')?)';
var open_tag    = '<[A-Z][A-Za-z0-9\\-]*' + attribute + '*\\s*\\/>';

// var close_tag   = '<\\/[A-Z][A-Za-z0-9\\-]*\\s*>';

export const REACT_TAG_RE = new RegExp('^(?:' + open_tag + ')');
export const REACT_OPEN_CLOSE_TAG_RE = /^<([A-Z][A-Za-z0-9]*)(?:.|[\n\f\r\t\v])*?<\/\1>/;