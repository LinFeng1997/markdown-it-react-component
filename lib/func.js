export function filterXSS() {
  return {
    visitor: {
      Identifier(path) {
        if (path.isIdentifier({ name: 'constructor' })) {
          // console.warn('constructor xss');
          path.node.name = 'x';
        }
        else if (typeof window !== 'undefined' && window[path.node.name]) {
          path.node.name = 'x';
        }
      },
      // Function traverse
      FunctionExpression(path) {
        path.traverse({
          ThisExpression(path) {
            // console.warn('FunctionExpression xss');
            path.remove();
          },
        });
      },
      FunctionDeclaration(path) {
        const funcName = path.node.id.name;

        if (funcName !== 'YMGGKLK') {
          path.traverse({
            ThisExpression(path) {
              // console.warn('FunctionDeclaration xss');
              path.remove();
            },
          });
        }
      },
      ThisExpression() {
        // console.log('this');
      },
      JSXIdentifier(path) {
        if (path.node.name === 'dangerouslySetInnerHTML'){
          path.node.name = 'sanitizition';
        }
      }
    }
  };
}

// compile javascript code with sandbox
export function compileCode(code) {
  code = 'with (sandbox) {' + code + '}';
  const fn = new Function('sandbox', code);
  return (sandbox) => {
    const proxy = new Proxy(sandbox, {
      has() {
        return true; // cheet prop
      },
      get(target, key, receiver) {
        // protect escape
        if (key === Symbol.unscopables) {
          return undefined;
        }
        return Reflect.get(target, key, receiver);
      }
    });
    return fn(proxy);
  };
}

// render react code to dom
export function renderReact(wrapperId, reactCode, sandbox) {
  const __container = document.getElementById(wrapperId);
  if (!__container) {
    return;
  }

  const newSandbox = {
    ...sandbox,
    __container,
  };
  const func = compileCode(reactCode);
  func(newSandbox);
  // style resume
  __container.style.opacity = 1;
}