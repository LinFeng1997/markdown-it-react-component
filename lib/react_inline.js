var attr_name     = '[a-zA-Z_:][a-zA-Z0-9:._-]*';
var unquoted      = '[^"\'=<>`\\x00-\\x20]+';
var single_quoted = "'[^']*'";
var double_quoted = '"[^"]*"';

var attr_value  = '(?:' + unquoted + '|' + single_quoted + '|' + double_quoted + ')';
var attribute   = '(?:\\s+' + attr_name + '(?:\\s*=\\s*' + attr_value + ')?)';
var open_tag    = '<[A-Z][A-Za-z0-9\\-]*' + attribute + '*\\s*\\/>';

// var close_tag   = '<\\/[A-Z][A-Za-z0-9\\-]*\\s*>';

export const REACT_TAG_RE = new RegExp('^(?:' + open_tag + ')');

const REACT_OPEN_CLOSE_TAG_RE = /^<([A-Z][A-Za-z0-9]*)(.|\s)*<\/\1>/;
// const REACT_TAG_RE = /^<([A-Z][A-Za-z0-9]*)\/>$/;

function isLetter(ch) {
  /*eslint no-bitwise:0*/
  var lc = ch | 0x20; // to lower case
  return (lc >= 0x61/* a */) && (lc <= 0x7a/* z */);
}

function isFailedSecondCh(ch) {
  return ch !== 0x21/* ! */ &&
    ch !== 0x3F/* ? */ &&
    ch !== 0x2F/* / */ &&
    !isLetter(ch);
}

export function react_inline(state, silent) {
  let pos = state.pos;

  if (!state.md.options.html) { return false; }

  function isStart(){
    return state.src.charCodeAt(pos) === 0x3C/* < */ &&
      pos + 2 < state.posMax;
  }

  // Check start
  if (!isStart()) {
    return false;
  }

  // Quick fail on second char
  if (isFailedSecondCh(state.src.charCodeAt(pos + 1))) {
    return false;
  }

  let match = state.src.slice(pos).match(REACT_OPEN_CLOSE_TAG_RE) || state.src.slice(pos).match(REACT_TAG_RE);
  if (!match) { return false; }

  if (!silent) {
    let token         = state.push('react_inline', '', 0);
    token.content = state.src.slice(pos, pos + match[0].length);
  }
  state.pos += match[0].length;
  return true;
};