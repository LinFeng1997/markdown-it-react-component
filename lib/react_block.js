// React block
// An array of opening and corresponding closing sequences for react jsx tags,
// last argument defines whether it can terminate a paragraph or not
//
import { REACT_TAG_RE } from './constans';

var REACT_SEQUENCES = [
  // PascalCase Components
  [REACT_TAG_RE, /^$/, true],
  [/^<([A-Z][A-Za-z0-9.]*)/, /^$/, true, (p1) => {
    return new RegExp(`<?\/(${p1})?>\\s*$`);
  }]
  // [ /^<![A-Z]/,     />/,     true ],
  // [ new RegExp('^</?(' + block_names.join('|') + ')(?=(\\s|/?>|$))', 'i'), /^$/, true ],
  // [ new RegExp(HTML_OPEN_CLOSE_TAG_RE.source + '\\s*$'),  /^$/, false ]
];
var htmlReg = /<[a-z]+.*>/;

function getCloseTagReg(tag, regs) {
  let reg = null;
  if (tag && typeof regs[3] === 'function') {
    reg = regs[3](tag);
  }
  return reg;
}
function pushStack(stack,lineText) {
  let openTag = null;
  let i = 0;
  for (; i < REACT_SEQUENCES.length; i++) {
    const match = lineText.match(REACT_SEQUENCES[i][0]);
    openTag = match ? match[1] : null;
    const tagMatch = getCloseTagReg(openTag,REACT_SEQUENCES[i]);
    if (tagMatch) {
      const isInline = tagMatch.test(lineText);
      // <Timeline.Item>aaa</Timeline.Item>
      if (!isInline) {
        stack.push({type: 'open', str: lineText, tag: openTag});
      }
      break;
    }
  }
  return {openTag,i};
}
export default function react_block(state, startLine, endLine, silent) {
  var nextLine, token,
    pos = state.bMarks[startLine] + state.tShift[startLine],
    max = state.eMarks[startLine],
    stack = [];

  // if it's indented more than 3 spaces, it should be a code block
  if (state.sCount[startLine] - state.blkIndent >= 4) { return false; }

  if (!state.md.options.html) { return false; }

  if (state.src.charCodeAt(pos) !== 0x3C/* < */) { return false; }

  let lineText = state.src.slice(pos, max);

  // 入栈第一处
  let {openTag, i} = pushStack(stack, lineText);

  if (i === REACT_SEQUENCES.length) { return false; }

  if (silent) {
    // true if this sequence can be a terminator, false otherwise
    return REACT_SEQUENCES[i][2];
  }

  nextLine = startLine + 1;

  const endReg = getCloseTagReg(openTag,REACT_SEQUENCES[i]) || REACT_SEQUENCES[i][1];
  // If we are here - we detected HTML block.
  // Let's roll down till block end.
  if (!endReg.test(lineText)) {
    let innerEndReg = endReg;
    for (; nextLine < endLine; nextLine++) {
      if (state.sCount[nextLine] < state.blkIndent) { break; }

      pos = state.bMarks[nextLine] + state.tShift[nextLine];
      max = state.eMarks[nextLine];
      lineText = state.src.slice(pos, max);
      // 入栈第二处
      const {openTag: innerTag} = pushStack(stack, lineText);
      innerEndReg = getCloseTagReg(innerTag,REACT_SEQUENCES[i]) || innerEndReg;
      if(stack.length === 1){
        innerEndReg = endReg;
      }
      if ((innerEndReg.test(lineText)) && !htmlReg.test(lineText) && !innerTag) {
        // 出栈处,恢复现场
        stack.pop();
        if(stack.length === 0){
          if (lineText.length !== 0) { nextLine++; }
          break;
        }
      }
    }
  }

  state.line = nextLine;

  token         = state.push('react_block', '', 0);
  token.map     = [ startLine, nextLine ];
  token.content = state.getLines(startLine, nextLine, state.blkIndent, true);

  return true;
}
