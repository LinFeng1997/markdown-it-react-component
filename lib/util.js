import React from 'react'
import ReactDOM from 'react-dom'
import { babelTransformOptions, filterXSS } from './constans'
import * as babel from 'babel-standalone'

// use babel filter xss attack
babel.registerPlugin('filterXSS', filterXSS)

function transReactCode (code, babelOptions = {}) {
  const rst = babel.transform(code, {...babelTransformOptions, ...babelOptions})

  return rst.code
}

export function replacerReact (code, wrapperId, components, babelOptions) {
  const reactCode = `function T(props) {
                    ${code}
                  }
                  ReactDOM.hydrate(T(),__container)
                  `

  const rst = transReactCode(reactCode, babelOptions)
  setTimeout(renderReact.bind(null, wrapperId, rst, components), 0)
}

export function renderReact (wrapperId, reactCode, components) {
  const __container = document.getElementById(wrapperId)
  if (!__container) {
    return
  }
  const sandbox = {
    React,
    ReactDOM,
    ...components,
    __container,
  }
  const func = compileCode(reactCode)
  func(sandbox)
  // style resume
  __container.style.opacity = 1
}

export function compileCode (code) {
  code = 'with (sandbox) {' + code + '}'
  const fn = new Function('sandbox', code)
  return (sandbox) => {
    const proxy = new Proxy(sandbox, {
      has () {
        return true //
      },
      get (target, key, receiver) {
        //
        if (key === Symbol.unscopables) {
          return undefined
        }
        return Reflect.get(target, key, receiver)
      }
    })
    return fn(proxy)
  }
}