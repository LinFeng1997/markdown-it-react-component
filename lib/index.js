import React from 'react';
import ReactDOM from 'react-dom';
import ReactDOMServer from 'react-dom/server';
import * as babel from 'babel-standalone';
import { babelTransformOptions } from './constans';
import { compileCode, filterXSS,renderReact } from './func';

const NOOP = () => {
};
export default class Replacer {
  constructor(options = {}) {
    if (options.sandbox && typeof options.sandbox !== 'object') {
      throw Error('sandbox must be object!');
    }
    if (options.babelInit && typeof options.babelInit !== 'function') {
      throw Error('babelInit must be function!');
    }
    this._init(options);
  }

  _init({
    babelInit = NOOP,
    sandbox = {},
    babelOptions,
    enableSSR = true
  }) {
    babel.registerPlugin('filterXSS', filterXSS);
    babelInit(babel);
    this.babel = babel;
    this.babelOptions = {
      ...babelTransformOptions,
      ...babelOptions
    };
    this.sandbox = {
      React,
      ReactDOM,
      NULL: Object.create(null),
      ReactDOMServer,
      ...sandbox
    };
    this.enableSSR = enableSSR;
  }

  _replace({
    str,
    wrapperId,
    lang,
  }) {
    if (lang === 'mixin-react') {
      return this._replacerReact(str, wrapperId);
    }
    return str;
  }

  // get html of react code
  _replacerReact(code, wrapperId) {
    const reactCode = `function YMGGKLK(props) { ${code} }
                  ReactDOM.hydrate(YMGGKLK.call(NULL),__container)
                  `;

    const finalCode  = this._transReactCode(reactCode);
    const html = this.enableSSR ? this._codeToHtml(wrapperId, code) : '<div></div>';

    setTimeout(renderReact.bind(null, wrapperId, finalCode, this.sandbox), 0);

    return html;
  }

  // pre-render HTML
  _codeToHtml(wrapperId, code) {
    const reactCode = `function YMGGKLK(props) { ${code} }
                  Html["${wrapperId}"] = ReactDOMServer.renderToStaticMarkup(YMGGKLK.call(NULL))
                  `;
    const Html = {};
    const newSandbox = {
      ...this.sandbox,
      Html
    };
    const func = compileCode(this._transReactCode(reactCode));
    func(newSandbox);
    return Html[wrapperId];
  }

  // use babel translation react code
  _transReactCode(code) {
    const rst = this.babel.transform(code, this.babelOptions);

    return rst.code;
  }

  getHtml(wrapperId,str,lang = 'mixin-react') {
    const html = this._replace({
      str,
      wrapperId,
      lang,
    });
    return html;
  }
}