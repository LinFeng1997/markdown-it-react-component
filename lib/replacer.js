import { replacerReact } from './util'

// Todo:babel options
export default function replacer ({str, wrapperId, lang, components = {}}) {
  console.log(1)
  if (lang === 'mixin-react') {
    console.log(2)
    return replacerReact(str, wrapperId, components)
  }
  return str
}

