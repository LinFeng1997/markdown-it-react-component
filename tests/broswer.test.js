// 前端
// dom
import MarkdownIt from 'markdown-it';
import Enzyme,{ render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { SupportReactComponent } from '../index';
import React from 'react';
import { getSimpleInput,wrapperId } from './util';
import * as funcs from '../lib/func';

funcs.getWrapperId = () => wrapperId;

Enzyme.configure({ adapter: new Adapter() })
const Hello = (props) => <div>{props.text} world!</div>;
const md = new MarkdownIt({
    breaks: true,
    html: true
}).use(SupportReactComponent, {
    sandbox: {
        React,
        Array: {
            isArray: Array.isArray
        },
        Hello
    }
});
describe('render', () => {
    it('should support render', (done) => {
        document.body.innerHTML = `<div id="${wrapperId}"></div>`
        const html = md.render(getSimpleInput('component'));
        const wrapper = render(
            <div dangerouslySetInnerHTML={{ __html: html }}></div>,
        );
        expect(wrapper).toMatchSnapshot();
        setTimeout(()=>{
            done();
        },100)
    });
});