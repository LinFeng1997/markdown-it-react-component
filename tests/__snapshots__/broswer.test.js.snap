// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should support render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rc",
          "id": "wrapperId",
          "style": "opacity: 0",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "hello world!",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {},
              "children": Array [],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "data": "
",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "text",
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Object {
            "attribs": Object {},
            "children": Array [],
            "name": "p",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "data": "
",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "text",
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "hello world!",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Object {
            "data": "
",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "hello world!",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "data": "
",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "text",
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "id": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "id": undefined,
          "style": undefined,
        },
      },
      Object {
        "data": "
",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "rc",
            "id": "wrapperId",
            "style": "opacity: 0",
          },
          "children": Array [
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "hello world!",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {},
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "data": "
",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "text",
                },
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Object {
              "attribs": Object {},
              "children": Array [],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "data": "
",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "text",
              },
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "hello world!",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Object {
              "data": "
",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "hello world!",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "id": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "id": undefined,
            "style": undefined,
          },
        },
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
